### PR TITLE
[SPARK-42520][CONNECT] Support basic Window API in Scala client

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Support Window orderby, partitionby, rowsbetween/rangebetween.

### Why are the changes needed?

API coverage

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala [hunk 1]
```diff
@@ -1233,6 +1233,22 @@ class Column private[sql] (private[sql] val expr: proto.Expression) extends Logg
...
+   *     sum("price").over(w.rangeBetween(Window.unboundedPreceding, 2)),
+   *     avg("price").over(w.rowsBetween(Window.currentRow, 4))
+   *   )
+   * }}}
+   *
+   * @group expr_ops
+   * @since 3.4.0
+   */
+  def over(window: expressions.WindowSpec): Column = window.withAggregate(this)
 }
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1252]: Can you also add the empty variant?

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1269]: done

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Defines the ordering columns in a [[WindowSpec]].
+   * @since 3.4.0
+   */
+  @scala.annotation.varargs
+  def orderBy(cols: Column*): WindowSpec = {
+    val sortOrder: Seq[proto.Expression.SortOrder] = cols.map { col =>
+      col.expr match {
+        case expr if expr.hasSortOrder =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala:74]: Can you use `Column.sortOrder` here?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+          proto.Expression.Window.WindowFrame.FrameType.FRAME_TYPE_RANGE,
+          start,
+          end)))
+  }
+
+  /**
+   * Converts this [[WindowSpec]] into a [[Column]] with an aggregate expression.
+   */
+  private[sql] def withAggregate(aggregate: Column): Column = {
+    Column({ builder =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala:204]: Nit you can drop the Parenthesis if you use curly brackets.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  private[sql] def toWindowFrame(
+      frameType: proto.Expression.Window.WindowFrame.FrameType,
+      start: Long,
+      end: Long): proto.Expression.Window.WindowFrame = {
+    val windowFrameBuilder = proto.Expression.Window.WindowFrame.newBuilder()
+    windowFrameBuilder.setFrameType(frameType)
+    start match {
+      case 0 => windowFrameBuilder.getLowerBuilder.setCurrentRow(true)
+      case Long.MinValue => windowFrameBuilder.getLowerBuilder.setUnbounded(true)
+      case x if Int.MinValue <= x && x <= Int.MaxValue =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala:224]: Why do we need to stay within the Int range? The old client allows the long range.

**Review by hvanhovell — COMMENTED** [review]
Looks good. A few small comments.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala (hunk 2 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala:74]: Can you share a bit more on how to use `Column,sortOrder` here?

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala:74]: It is a function that does the same thing you are doing here.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala (hunk 4 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala:224]: good catch. updated.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4!

**Comment by LuciferYang** [comment]
Good work, late LGTM

**Pull request closed without merging**